modules/dhcp_common/free_ips.rb

```diff
@@ -0,0 +1,133 @@
+require 'concurrent'
+require 'logger'
+
+module Proxy::DHCP
+  class FreeIps
+    include ::Proxy::Log
+
+    DEFAULT_ADDRESS_KEEP_TIME = 30*60 # 30 mins
+    DEFAULT_CLEANUP_INTERVAL = 60 # 1 min
+
+    attr_reader :allocated_ips, :allocation_timestamps, :m
+
+    def initialize
+      @allocated_ips = []
+      @allocation_timestamps = []
+      @m = Monitor.new
+    end
+
+    # must be called from under the monitor
+    def mark_ip_as_allocated(ip_address)
+      @allocated_ips.push(ip_address)
+      @allocation_timestamps.push(time_now + DEFAULT_ADDRESS_KEEP_TIME)
+    end
+
+    def clean_up_allocated_ips
+      @m.synchronize do
+        loop do
+          logger.debug("Starting allocated ip addresses cleanup pass...")
+          break if @allocation_timestamps.first.nil? || @allocation_timestamps.first > time_now
+          @allocation_timestamps.shift
+          ip = @allocated_ips.shift
+          logger.debug("#{ip} marked as free.")
+        end
+      end
+    end
+
+    def start
+      logger.info("Starting allocated ip address maintenance (used by unused_ip call).")
+      @timer_task = Concurrent::TimerTask.new(:execution_interval => DEFAULT_CLEANUP_INTERVAL) { clean_up_allocated_ips }
+      @timer_task.execute
+    end
+
+    def stop
+      @timer_task.shutdown unless @timer_task.nil?
+    end
+
+    def time_now
+      Time.now.to_i
+    end
+
+    def find_free_ip(subnet_address, possible_ips, records)
+      @m.synchronize do
+        free_ips = possible_ips - @allocated_ips - records.collect{|record| record.ip}
+        if free_ips.empty?
+          logger.warn "No free IPs in #{subnet_address}."
+          return nil
+        end
+        begin
+          free_ips.shuffle.each do |ip| # shuffle available ip addresses to reduce collisions after proxy restart
+            logger.debug "Searching for free IP - pinging #{ip}."
+            if tcp_pingable?(ip) || icmp_pingable?(ip)
+              mark_ip_as_allocated(ip)
+              logger.debug "Found a pingable IP(#{ip}) address which does not have a Proxy::DHCP record."
+            else
+              logger.debug "Found free IP #{ip} out of a total of #{free_ips.size} free IPs."
+              mark_ip_as_allocated(ip)
+              return ip
+            end
+          end
+        rescue Exception => e
+          logger.debug e.message
```

Comment:
Never swallow exceptions into debug level, please print with error level. Debug is ideal target for stacktrace.